pkg/sql/jobs/registry.go

```diff
@@ -54,3 +94,202 @@ func (r *Registry) LoadJob(ctx context.Context, jobID int64) (*Job, error) {
 	}
 	return j, nil
 }
+
+// DefaultCancelInterval is a reasonable interval at which to poll this node
+// for liveness failures and cancel running jobs.
+const DefaultCancelInterval = base.DefaultHeartbeatInterval
+
+// DefaultAdoptInterval is a reasonable interval at which to poll system.jobs
+// for jobs with expired leases.
+//
+// DefaultAdoptInterval is mutable for testing. NB: Updates to this value after
+// Registry.Start has been called will not have any effect.
+var DefaultAdoptInterval = 30 * time.Second
+
+// Start polls the current node for liveness failures and cancels all registered
+// jobs if it observes a failure.
+func (r *Registry) Start(
+	stopper *stop.Stopper, nl nodeLiveness, cancelInterval, adoptInterval time.Duration,
+) error {
+	stopper.RunWorker(context.Background(), func(ctx context.Context) {
+		for {
+			select {
+			case <-time.After(cancelInterval):
+				r.maybeCancelJobs(ctx, nl)
+			case <-stopper.ShouldStop():
+				return
+			}
+		}
+	})
+
+	stopper.RunWorker(context.Background(), func(ctx context.Context) {
+		for {
+			select {
+			case <-time.After(adoptInterval):
+				if err := r.maybeAdoptJob(ctx, nl); err != nil {
+					log.Errorf(ctx, "error while adopting jobs: %+v", err)
+				}
+			case <-stopper.ShouldStop():
+				return
+			}
+		}
+	})
+	return nil
+}
+
+func (r *Registry) maybeCancelJobs(ctx context.Context, nl nodeLiveness) {
+	liveness, err := nl.Self()
+	if err != nil {
+		log.Warningf(ctx, "unable to get node liveness: %s", err)
+		// Conservatively assume our lease has expired. Abort all jobs.
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		r.cancelAll(ctx)
+		return
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// TODO(benesch): cancel only jobs from previous epochs. This logic is correct
+	// but too aggressive, as it might cancel jobs created immediately after a
```

Comment:
It also might cancel jobs even if the node's liveness never expired. For example:

1. The node sends a heartbeat to update its liveness expiration
1. The heartbeat applies before the node's liveness expires
1. The response to the heartbeat takes a while to get back to the node, so its locally cached liveness is a little out of date and indicates that the node is no longer live.

In these cases, we will cancel things here but we don't actually need to. This can also be left as a TODO, but it's another thing that could be improved.